pkg/resmgr/lib/memory/types.go

```diff
@@ -0,0 +1,318 @@
+}
+
+// Sysfs returns the sysfs memory type for the given memory type.
+func (t Type) Sysfs() sysfs.MemoryType {
+	if sysType, ok := typeToSys[t]; ok {
+		return sysType
+	}
+
+	panic(fmt.Errorf("unknown libmem memory type %d", t))
+}
+
+// ParseType parses the given string into a memory type.
+func ParseType(str string) (Type, error) {
+	if t, ok := stringToType[strings.ToUpper(str)]; ok {
+		return t, nil
+	}
+
+	return 0, fmt.Errorf("%w: %q", ErrInvalidType, str)
+}
+
+// MustParseType parses the given string into a memory type.
+// It panicks on failure.
+func MustParseType(str string) Type {
+	t, err := ParseType(str)
+	if err == nil {
+		return t
+	}
+
+	panic(err)
+}
+
+// Mask returns the memory TypeMask for the memory type.
+func (t Type) Mask() TypeMask {
+	return TypeMask(1 << t)
+}
+
+// IsValid returns true if the memory type is valid/known.
+func (t Type) IsValid() bool {
+	_, ok := typeToSys[t]
+	return ok
+}
+
+// String returns a string representation of the memory type.
+func (t Type) String() string {
+	if str, ok := typeToString[t]; ok {
+		return str
+	}
+
+	return fmt.Sprintf("%%!(libmem:Bad-Type %d)", t)
+}
+
+// MarshalJSON is the json.Marshaller for Type.
+func (t Type) MarshalJSON() ([]byte, error) {
+	return json.Marshal(t.String())
+}
+
+// UnmarshalJSON is the json.Unmarshaller for Type.
+func (t *Type) UnmarshalJSON(data []byte) error {
+	i := 0
+	if err := json.Unmarshal(data, &i); err == nil {
+		if _, ok := typeToString[Type(i)]; ok {
+			*t = Type(i)
+			return nil
+		}
+		return fmt.Errorf("%w: %d", ErrInvalidType, i)
+	}
+
+	str := ""
+	err := json.Unmarshal(data, &str)
+	if err != nil {
+		return fmt.Errorf("%w: %w", ErrInvalidType, err)
+	}
+
+	stt, ok := stringToType[strings.ToUpper(str)]
+	if !ok {
+		return fmt.Errorf("%w: %s", ErrInvalidType, str)
+	}
+
+	*t = stt
+	return nil
+}
+
+// TypeMask represents a set of memory types as a bit mask.
+type TypeMask int
+
+const (
+	TypeMaskDRAM TypeMask = 1 << TypeDRAM          // ordinary DRAM
+	TypeMaskPMEM TypeMask = 1 << TypePMEM          // 'persistent' memory
+	TypeMaskHBM  TypeMask = 1 << TypeHBM           // high-bandwidth memory
+	TypeMaskAll  TypeMask = (TypeMaskHBM << 1) - 1 // all types of memory
+)
+
+var (
+	typeMaskToString map[TypeMask]string
+)
+
+// NewTypeMask returns a TypeMask containing the given memory types.
+func NewTypeMask(types ...Type) TypeMask {
+	m := TypeMask(0)
+	for _, t := range types {
+		m |= (1 << t)
+	}
+	return m & TypeMaskAll
+}
+
+// NewTypeMaskForSysfs returns a TypeMask containing th given sysfs memory types.
+func NewTypeMaskForSysfs(sysTypes ...sysfs.MemoryType) TypeMask {
+	m := TypeMask(0)
+	for _, st := range sysTypes {
+		m |= (1 << TypeForSysfs(st))
+	}
+	return m & TypeMaskAll
+}
+
+// ParseTypeMask parses the given string into a TypeMask.
+func ParseTypeMask(str string) (TypeMask, error) {
+	m := TypeMask(0)
+	for _, s := range strings.Split(str, ",") {
+		t, err := ParseType(s)
```

Comment:
Minor details... Do you think ParseTypeMask could allow whitespace around the mask string and commas? I have stumbled into this a couple of time in our current `memory-type` annotation parser, as it's quite natural to type annotations with spaces: `memory-types...: hbm, dram`.

I'd like to use this parser in refactoring the annotation code and making it work in both balloons and topology-aware.